nav2_assisted_teleop/src/nav2_assisted_teleop.cpp

```diff
@@ -0,0 +1,311 @@
+  {
+    RCLCPP_INFO(logger_, "Cleaning up assisted_teleop_server");
+
+    vel_pub_->on_activate();
+    action_server_->activate();
+
+    return nav2_util::CallbackReturn::SUCCESS;
+  }
+
+  void AssistedTeleopServer::adjustVelocityIfInvalid()
+  {
+    vel_sub_ = create_subscription<geometry_msgs::msg::Twist>(
+    input_vel_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleopServer::inputVelocityCallback,
+      this, std::placeholders::_1));
+
+    joy_sub_ = create_subscription<sensor_msgs::msg::Joy>(
+    joystick_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleopServer::joyCallback,
+      this, std::placeholders::_1));
+
+    auto start_time = steady_clock_.now();
+
+    auto result = std::make_shared<Action::Result>();
+    rclcpp::WallRate loop_rate(10);
+    while (rclcpp::ok())
+    {
+      RCLCPP_INFO(logger_, "Running");
+      // user states that teleop was successful
+      if( joy_.buttons.size() > 0 && joy_.buttons[0] > 0)
+      {
+        vel_pub_->publish(geometry_msgs::msg::Twist());
+        vel_sub_.reset();
+        joy_sub_.reset();
+        result->total_elapsed_time = steady_clock_.now() - start_time;
+        action_server_->succeeded_current(result);
+        return;
+      }
+
+      // user states that teleop failed
+      if( joy_.buttons.size() > 1 && joy_.buttons[1] > 0)
+      {
+        vel_pub_->publish(geometry_msgs::msg::Twist());
+        vel_sub_.reset();
+        joy_sub_.reset();
+        result->total_elapsed_time = steady_clock_.now() - start_time;
+        action_server_->terminate_all(result);
+        return;
+      }
+
+      // cancel request
+      if(action_server_->is_cancel_requested())
+      {
+        vel_pub_->publish(geometry_msgs::msg::Twist());
+        vel_sub_.reset();
+        joy_sub_.reset();
+        result->total_elapsed_time = steady_clock_.now() -start_time;
+        action_server_->terminate_all(result);
+        return;
+      }
+
+      // preempt request
+      if (action_server_->is_preempt_requested()) {
+        vel_pub_->publish(geometry_msgs::msg::Twist());
+        vel_sub_.reset();
+        joy_sub_.reset();
+        result->total_elapsed_time = steady_clock_.now() - start_time;
+        action_server_->terminate_current(result);
+        return;
+      }
+
+      geometry_msgs::msg::Twist cmd_vel;
+
+      cmd_vel = assisted_teleop_->computeVelocity(input_twist_);
+
+      vel_pub_->publish(std::move(cmd_vel));
+      loop_rate.sleep();
+    }
+  }
+
+  void AssistedTeleopServer::inputVelocityCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
+  {
+    RCLCPP_INFO(logger_, "Hit input Vel callback");
+
+    input_twist_.linear.x = msg->linear.x;
+    input_twist_.linear.y = msg->linear.y;
+    input_twist_.angular.z = msg->linear.z;
+  }
+
+  void AssistedTeleopServer::joyCallback(const sensor_msgs::msg::Joy msg)
+  {
+    RCLCPP_INFO(logger_, "Hit joy callback");
+    joy_.buttons = msg.buttons;
```

Comment:
Ditto, why not just store everything `joy_ = *msg` (with shared pointer). Also, `input_joy_` might be better so that its analog to the twist input

Reply:
Probably should be removed if we can't make any general rules about joys